File: week01/pset1/greedy.c
#include <cs50.h>
#include <stdio.h>

int main(void)
{
	// first deal with input
	printf("O hai! ");
	float raw_input;
	do
	{
		printf("How much change is owed?\n");
		raw_input = GetFloat();
	}
	while(raw_input < 0.0);

    /*
    ** now it gets UGLY
    ** this is to deal with 4.2 * 100 = 419.999
    ** SERISLY, WAT DA FUK, C?!
    */
    int thousands = (int)(raw_input * 1000);
    int cents = thousands / 10;
    int remainder = thousands % 10;
    if (remainder == 9) cents++;

	// now the algo itself
	// while there is change
	int result = 0;
	while(cents)
	{
		if (cents >= 25)
		{
			cents -= 25;
			result++;
		} else if (cents >= 10)
		{
			cents -= 10;
			result++;
		} else if (cents >= 5)
		{
			cents -= 5;
			result++;
		} else if (cents >= 1)
		{
			cents -= 1;
			result++;
		}
	}

	// produce result to the user on screen
	printf("%d\n", result);
}

File: week08/pset6/greedy.py
import cs50
from math import floor

def press_lever(value, cents, result):
    while cents >= value:
        cents -= value
        result += 1
    return cents, result

def check_coins(change):
    result = 0
    cents = int(floor(change * 100)) # truncates if values smaller than cents are provided $4.1111 -> 411 cents
    for value in [25, 10, 5, 1]:
        cents, result = press_lever(value, cents, result)
    return result

def main():
    print("O hai! How much change is owed?")
    while True:
        change = cs50.get_float()
        if change >= 0:
            break
        print("How much change is owed?")

    result = check_coins(change)
    print(result)
    return 1
    
if __name__ == "__main__":
    main()